### PR TITLE
Attempt to "Fix" OpenBSD Build

### PR DESCRIPTION
**Pull request by time-killer-games**
I will need someone to do additional research and make sure this is correct as I'd like to not be responsible for further breakage to what is already borked.

### CODE REVIEW

File: openbsd/OpenBSDMachine.c [hunk 1]
```diff
@@ -114,7 +115,7 @@ Machine* Machine_new(UsersTable* usersTable, uid_t userId) {
 
    this->cpuSpeed = -1;
 
-   return this;
+   return (Machine*)this;
```
Comment (BenBE) [inline openbsd/OpenBSDMachine.c:118]:
```suggestion
   return super;
```

File: openbsd/OpenBSDMachine.c [hunk 2]
```diff
@@ -92,11 +92,12 @@ Machine* Machine_new(UsersTable* usersTable, uid_t userId) {
    char errbuf[_POSIX2_LINE_MAX];
 
    OpenBSDMachine* this = xCalloc(1, sizeof(OpenBSDMachine));
-   Machine* super = &this->super;
+   Machine* super = (Machine*)&this->super;
```
Comment (BenBE) [inline openbsd/OpenBSDMachine.c:95]: What's the error message for this? `this->super` should be of type `Machine` in `openbsd/OpenBSDMachine.h`.

File: openbsd/OpenBSDMachine.c [hunk 3]
```diff
@@ -92,11 +92,12 @@ Machine* Machine_new(UsersTable* usersTable, uid_t userId) {
...
 
    Machine_init(super, usersTable, userId);
 
-   OpenBSDProcessList_updateCPUcount(this);
+   // Commented this out temporarily to fix build error incorrectly:
+   //OpenBSDProcessList_updateCPUcount(this); // How do I fix it ???
```
Comment (BenBE) [inline openbsd/OpenBSDMachine.c:100]: There seems to be that some code got lost in 72235d8e098d9d79029dca65122605741e1aafad when @natoscott did some refactoring to separate the Machine code from the Process List stuff.

Have a look at that commit (and those nearby) and look for that routine (should have been in `OpenBSDProcessList.c`). The new location for this should be `OpenBSDMachine.c` following the intention of PR #1234 (cf. #1238, #1239, #1240).

File: openbsd/OpenBSDMachine.c [hunk 4]
```diff
@@ -129,7 +130,7 @@ void Machine_delete(Machine* super) {
 }
 
 static void OpenBSDMachine_scanMemoryInfo(OpenBSDMachine* this) {
-   Machine* host = &this->super;
+   Machine* super = (Machine*)&this->super;
```
Comment (BenBE) [inline openbsd/OpenBSDMachine.c:133]: There should be no cast necessary. Please provide build message.

File: openbsd/OpenBSDMachine.c [hunk 5]
```diff
@@ -139,18 +140,7 @@ static void OpenBSDMachine_scanMemoryInfo(OpenBSDMachine* this) {
...
-   // Taken from OpenBSD systat/iostat.c, top/machine.c and uvm_sysctl(9)
-   const int bcache_mib[] = { CTL_VFS, VFS_GENERIC, VFS_BCACHESTAT };
-   struct bcachestats bcstats;
-   size_t size_bcstats = sizeof(bcstats);
-
-   if (sysctl(bcache_mib, 3, &bcstats, &size_bcstats, NULL, 0) < 0) {
-      CRT_fatalError("cannot get vfs.bcachestat");
-   }
-
-   super->cachedMem = bcstats.numbufpages * this->pageSizeKB;
```
Comment (BenBE) [inline openbsd/OpenBSDMachine.c:153]: Why is this part of the code removed?

Comment (time-killer-games) [reply openbsd/OpenBSDMachine.c:153]: I'm not sure what purpose the cached memory served, as I wasn't seeing that printed to the terminal as it's own stat in specific, although it does effect the amount of used ram reported, which is part of what made it not match up with top(1).

File: openbsd/OpenBSDMachine.c (hunk 4 above)
Comment (time-killer-games) [reply openbsd/OpenBSDMachine.c:133]: iirc it was saying there was a cast needed, which is odd. I didn't think that was necessary in C.

I'll revert the casts and show you the errors I'm getting.

Comment (time-killer-games) [reply openbsd/OpenBSDMachine.c:133]: Ah, i see now; they were warnings, not errors. In that case I'll revert that.

**Comment by time-killer-games** [comment]
i clicked commit on that by mistake i meant to revert them all at once.

File: openbsd/OpenBSDMachine.c (hunk 3 above)
Comment (natoscott) [reply openbsd/OpenBSDMachine.c]: @time-killer-games @BenBE yep, that's a spot I've missed - please replace OpenBSDProcessList_updateCPUcount with OpenBSDMachine_updateCPUcount here.  Thanks!

File: openbsd/OpenBSDMachine.c [hunk 6]
```diff
@@ -273,7 +262,7 @@ static void OpenBSDMachine_scanCPUTime(OpenBSDMachine* this) {
 }
 
 void Machine_scan(Machine* super) {
-   OpenBSDMachine* this = (OpenBSDMachine*) super;
+   OpenBSDMachine* this = super;
```
Comment (time-killer-games) [inline openbsd/OpenBSDMachine.c:265]: I didn't know this cast was here initially. Let me know if you want me to add it back. Though removing it didn't mess up the build. I might redo this pr from scratch if i mess up too much.

Comment (BenBE) [reply openbsd/OpenBSDMachine.c:265]: The rule usually is casting from generic to specific (platform) type; no casting from platform type to generic (use `->super` instead).

Also: Feel free to rebase/squash the commits to your hearts content, until they look clean.

File: openbsd/OpenBSDMachine.c (hunk 4 above)
Comment (BenBE) [reply openbsd/OpenBSDMachine.c]: Can you post the warnings you are getting? Just to check if there's anything funny going on.

File: openbsd/OpenBSDMachine.c (hunk 5 above)
Comment (BenBE) [reply openbsd/OpenBSDMachine.c:153]: There'S a later place in the code that calculates the used memory shown in the Memory meter. Also the cached memory (i.e. file system cache) is not directly printed there, but shown as blue (file buffers/shared memory) or brown (volatile file system cache).

htop tries to track some more resources separately, thus things might seem unintuitive at first.

File: openbsd/OpenBSDMachine.c (hunk 4 above)
Comment (time-killer-games) [reply openbsd/OpenBSDMachine.c]: It's the same warning that can be found in other sources I did not modify, such as this one:
```
OpenBSDProcessList.c:41:11: warning: incompatible pointer types returning 'OpenBSDProcessList *' (aka 'struct OpenBSDProcessList_ *') from a function with result type 'ProcessList *' (aka 'struct ProcessList_ *') [-Wincompatible-pointer-types]
```

There seems to be 3 other instances of this warning:
```
OpenBSDMachine.c:117:11: warning: incompatible pointer types returning 'OpenBSDMachine *' (aka 'struct OpenBSDMachine_ *') from a function with result type 'Machine *' (aka 'struct Machine_ *') [-Wincompatible-pointer-types]
```

```
OpenBSDMachine.c:121:20: warning: incompatible pointer types initializing 'OpenBSDMachine *' (aka 'struct OpenBSDMachine_ *') with an expression of type 'Machine *' (aka 'struct Machine_ *') [-Wincompatible-pointer-types]
```

```
OpenBSDMachine.c:265:20: warning: incompatible pointer types initializing 'OpenBSDMachine *' (aka 'struct OpenBSDMachine_ *') with an expression of type 'Machine *' (aka 'struct Machine_ *') [-Wincompatible-pointer-types]
```

File: openbsd/OpenBSDMachine.c (hunk 5 above)
Comment (time-killer-games) [reply openbsd/OpenBSDMachine.c:153]: So do you consider the used/free memory not matching top(1) to be intentional? If not, we will need to fix that, otherwise, this code can be put back and not much else needs changing at this point.

File: openbsd/OpenBSDMachine.c (hunk 6 above)
Comment (time-killer-games) [reply openbsd/OpenBSDMachine.c:265]: After I'm done making all my changes locally to how we want the final result to look, i will create a new pr, close this one, and break it up into a couple commits, ones that have code changes related to one another in how it is broken up, without the excess in extra commits. Is that ok?

In the meantime, feel free to let me know how you want the commits broken up here.

File: openbsd/OpenBSDMachine.c (hunk 5 above)
Comment (BenBE) [reply openbsd/OpenBSDMachine.c:153]: The intended display is used+compressed+buffers+shared for the used value in the memory bar; the cached value is ignored intentionally, as that's memory that can usually be reclaimed by the kernel if needed. And although cached resides in RAM, it's not usually considered as "used" for htop due to this volatile nature.

used: Memory used by programs, libraries, data
compressed: Memory compressed by te kernel to avoid swapping (e.g. zram)
buffers: File buffers & dirty pages for I/O
shared: Shared memory segments, memory used e.g. by tmpfs
cached: volatile file system cache (e.g. recently read data to optimize disk access)

Example from Linux:
image
Although there's practically no memory unused by Linux, the meter declares the more reasonable value of ~20GiB of memory that cannot easily be dropped, but would need to be placed somewhere else if some new program wanted to use too much memory.

File: openbsd/OpenBSDMachine.c (hunk 6 above)
Comment (BenBE) [reply openbsd/OpenBSDMachine.c]: Please instead of opening a new PR, just split the changes into commits as appropriate and push them (`--force-with-lease`) on the current branch.

Regarding splitting the commits:
- One commit (or more, if there are vastly distinct build issues) for compile issues
- One commit fixing the memory detection (see my other comment regarding that)

Apart from that, providing descriptive commit messages and following the information outlined in the style guide is much appreciated.

File: openbsd/OpenBSDMachine.c [hunk 7]
```diff
@@ -129,7 +129,7 @@ void Machine_delete(Machine* super) {
 }
 
 static void OpenBSDMachine_scanMemoryInfo(OpenBSDMachine* this) {
-   Machine* host = &this->super;
+   Machine* super = &this->super;
```
Comment (BenBE) [inline openbsd/OpenBSDMachine.c:132]: @natoscott Quick question re `host` vs `super`: Did we do a consistent rename of this previously and just miss that here? Using `super` is the "smaler" patch, but I have the feeling as if we were trying to go the direction of `host`. Fine either way, just consistent across platforms preferred.

Comment (natoscott) [reply openbsd/OpenBSDMachine.c:132]: I suspect it was just missed here - in situations where it is the super, we've used super - there's plenty of cases where its just a pointer field in a struct, or local variable, though and in those cases we use host.

File: openbsd/OpenBSDMachine.c (hunk 5 above)
Comment (time-killer-games) [reply openbsd/OpenBSDMachine.c:153]: I do want to make this a different pull request so this doesn't get conflated, but I think it would be good if NetBSD used code which was more in line with what OpenBSD is doing with the sysctl(), VM_UVMEXP, and swapctl(). NetBSD could be using near identical code for these things to reduce needed maintenance, but instead it is using sysctl() and the VM_UVMEXP2 struct, which is like the VM_UVMEXP struct but also covers what OpenBSD is doing to get swap related info with swapctl(). We'll cross that bridge when we get there, perhaps I'll link to this comment from my new pull request once this pull request is done.

**Comment by natoscott** [comment]
@time-killer-games I've merged the build-fix part of this PR, however lots of other changes have also been made in the meantime.  Could you see if we have a clean OpenBSD build now please?  Any issues, please let me know - thanks!

**Comment by time-killer-games** [comment]
> @time-killer-games I've merged the build-fix part of this PR, however lots of other changes have also been made in the meantime. Could you see if we have a clean OpenBSD build now please? Any issues, please let me know - thanks!

Without going too far into details my probation officer has taken all my devices except my phone. I don't when she will let me have my stuff again, but hopefully soon so i can do this for you. I apologize for the inconvenience.

**Comment by BenBE** [comment]
What's the status of this?

**Comment by BenBE** [comment]
@fraggerfox What about cross-checking this PR for #1360? They seem to address similar issues.

**Comment by time-killer-games** [comment]
> What's the status of this?

I'm in the position to work on this again, currently working on something unrelated, but wanted to let you know i got your message and sorry for the million years you had to wait. Will be working on this before the night is over.

**Comment by time-killer-games** [comment]
@BenBE do you and the other primary htop developers have a slack or discord account? I'd like to discuss what we are going to do in this pull request at realtime off-github so this pull request doesn't get flooded with comments or gets hard to follow. I'm ready to do what i can to help as we speak.

File: openbsd/OpenBSDMachine.c (hunk 7 above)
Comment (time-killer-games) [reply openbsd/OpenBSDMachine.c:132]: @natoscott @BenBE should this be changed back to `host` then, I am guessing?

**Comment by time-killer-games** [comment]
> @time-killer-games I've merged the build-fix part of this PR, however lots of other changes have also been made in the meantime. Could you see if we have a clean OpenBSD build now please? Any issues, please let me know - thanks!

I am going to attempt to resolve conflicts and we will go from there.

**Comment by time-killer-games** [comment]
Alright, tried building in the current state, and got these errors:

> openbsd/OpenBSDProcessTable.c:48:55: error: use of undeclared identifier 'super'
>    OpenBSDProcessTable* this = (OpenBSDProcessTable*) super;
>                                                       ^
> openbsd/OpenBSDProcessTable.c:47:34: warning: unused parameter 'cast' [-Wunused-parameter]
> void ProcessTable_delete(Object* cast) {
>                                  ^
> openbsd/OpenBSDProcessTable.c:133:32: error: no member named 'host' in 'struct ProcessTable_'
>    Machine* host = this->super.host;
>                    ~~~~~~~~~~~ ^
> 2 warnings and 2 errors generated.
> *** Error 1 in . (Makefile:1549 'openbsd/OpenBSDProcessTable.o')
> *** Error 2 in /home/openbsd/htop (Makefile:1095 'all')

File: openbsd/OpenBSDMachine.c (hunk 7 above)
Comment (BenBE) [reply openbsd/OpenBSDMachine.c]: Given that it's inside the `OpenBSDMachine_` namespace and we are pulling this from `&this->super` I'd tend to go with `super` for this variable. This note was mainly put as a question to @natoscott given that he did the PR that caused most of this renaming.

**Comment by BenBE** [comment]
The first error is with the cast needing `cast` instead of `super`. This is given that you get `Object *` (any parent class) instead of the direct parent (which would be named `super`).

This second one is inheritance-related as that field in question moved one more parent up in the hierarchy (`this->super.super.host`, unchecked). There will likely be a similar check in #1360. Also compare with other platforms like in e.g. `linux/LinuxProcessTable.c` and their handling of this kind of access.

**Comment by time-killer-games** [comment]
lol the needing of `cast` instead of `super` was rather obvious. I think it didn't occur to me at first because I never bothered looking up where that error was located in the actual code.

Fixing these things now. Will report back if everything worked.

**Comment by BenBE** [comment]
Please do not merge, but rather rebase your changes onto the latest commits from `main`.

**Comment by time-killer-games** [comment]
I think, unless I'm sadly mistaken, we're all set! :)

Screenshot_2024-01-05_12-17-39

**Comment by time-killer-games** [comment]
> Please do not merge, but rather rebase your changes onto the latest commits from `main`.

Taking care of it now.

**Comment by time-killer-games** [comment]
@BenBE is it ok if i open a new pull request and close this one, to reduce the number of commits down to one commit? I ask because I am having a bit of trouble rebasing, I'm not very good with git and I'm concerned I might mess this up.

If so, what should be the title and description of the commit and name of the branch?

**Comment by BenBE** [comment]
Preferably things should be resolved in this PR as it minimizes clutter in the issue tracker. If you really need to open a new one you should mark the new one to superseed this one, cross-link both issues accordingly so people can follow up in the new one.

I did a rebase+squash of your changes (as present on this PR). Please recheck. HTH.

In case you want to get a bit more familiar with how git works, there's a nice game called [Oh My Git](https://ohmygit.org/) that explains many of the basics and some advanced topics.

File: openbsd/OpenBSDMachine.c [hunk 8]
```diff
@@ -116,7 +116,7 @@ Machine* Machine_new(UsersTable* usersTable, uid_t userId) {
 
    this->cpuSpeed = -1;
 
-   return this;
+   return super;
```
Comment (time-killer-games) [inline openbsd/OpenBSDMachine.c:119]: @BenBE I'm not sure if this change is correct, let me know your thoughts. I can't remember why I changed `this` to `super`. This is likely my last concern before this gets merged, but I'm going to retest the pr on my end like you asked as we speak.

File: openbsd/OpenBSDMachine.c (hunk 5 above)
Comment (time-killer-games) [reply openbsd/OpenBSDMachine.c:155]: I almost forgot about this part; did you want me to add back the code logic I took out here? Again, this will make it not print the correct amount of ram and it will no longer align with what OpenBSD's stock `top` cli prints for ram.

**Comment by time-killer-games** [comment]
I think after you go over my two recent review comments with your thoughts, and we address those things, we should be all set. @BenBE

File: openbsd/OpenBSDMachine.c (hunk 8 above)
Comment (BenBE) [reply openbsd/OpenBSDMachine.c:119]: That change is correct, as `this` is `OpenBSDMachine*` and `super` is `Machine*`. If you look at the return type you see it match `Machine*`, thus `super`. With `&this->super == (Machine*)this`, that doesn't make much difference from a "value" perspective, but as `Machine* super = &this->super;` gives this value in a type-safe manner. I hope this explains this.

Also note: When I updated this PR earlier, I rechecked the various changes from a type perspective to make change.

File: openbsd/OpenBSDMachine.c (hunk 5 above)
Comment (BenBE) [reply openbsd/OpenBSDMachine.c:155]: Given that this change drops initializing the `super->cachedMem` member of the `Machine` structure, I'm inclined to have this block of code stay or be updated and follow the semantics used within `htop`.

The memory metrics in `htop` mostly fall into 4 categories (simplified):
1. used: RAM used by a program for working memory, code or other resources
2. shared: RAM shared between processes, like `shmem` or `mmap`'d files
3. buffers: RAM allocated for write-back buffers of the filesystem
4. caches: RAM used as filesystem (read) cache that could be dropped at any moment

Of those categories 1 and 2 are subject to potential swapping, while 4 is (mostly) never swapped. Category 3 is basically what you'd first write to disk, if you were to need some more free RAM.

This is much simplified, and differs slightly from what `top` displays regarding the handling of shared memory to accommodate for things like `tmpfs`/`ramfs` on Linux.

AFAICS, the code that was removed should have worked in this regard and if there's a need for that code to change I'd recommend to split out that change as a separate PR to separate build fixes from behavioral changes.

**Comment by time-killer-games** [comment]
trying to rebase, and i am only seeing the most recent one out of two commits total:

Screenshot_2024-01-05_21-24-54

Note sure what to make of that, but I can't mark the most recent one as "fixup" if there is no commit before that one. :/

**Comment by time-killer-games** [comment]
and... I just screwed up this pull request completely trying to rebase. Wonderful.

**Comment by time-killer-games** [comment]
Oh my god. This is the second time my rebase messed everything up.

**Comment by BenBE** [comment]
Superseeded by #1361 …
